src/native/libs/System.Security.Cryptography.Native/memory_debug.c

```diff
@@ -0,0 +1,309 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "opensslshim.h"
+#include "memory_debug.h"
+
+#include <pthread.h>
+#include <string.h>
+#include <assert.h>
+#include <stdatomic.h>
+
+//
+// OpenSSL memory tracking/debugging facilities.
+//
+// We can use CRYPTO_set_mem_functions to replace allocation routines in
+// OpenSSL. This allows us to prepend each allocated memory with a header that
+// contains the size and source of the allocation, which allows us to track how
+// much memory is OpenSSL consuming (not including malloc overhead).
+//
+// Additionally, if requested, we can track all allocations over a period of
+// time and present them to managed code for analysis (via reflection APIs).
+//
+// Given that there is an overhead associated with tracking, the feature is
+// gated behind the DOTNET_OPENSSL_MEMORY_DEBUG environment variable and should
+// not be enabled by default in production.
+//
+// To track all allocated objects in a given period, the allocated objects are
+// stringed in a circular, doubly-linked list. This allows us to do O(1)
+// insertion and deletion. All operations are done under lock to prevent data
+// corruption. To prevent lock contention over a single list, we maintain
+// multiple lists, each with its own lock and allocate them round-robin.
+//
+
+// helper macro for getting the pointer to containing type from a pointer to a member
+#define container_of(ptr, type, member) ((type*)((char*)(ptr) - offsetof(type, member)))
+
+typedef struct link_t
+{
+    struct link_t* next;
+    struct link_t* prev;
```

Comment:
The only thing I see `prev` used for is in the maintenance of the doubly-linked list.  Am I missing something where it's relevant that it's doubly-linked, or should this only have `next`?

Reply:
list being doubly linked allows us to remove from the middle of the list in O(1). During free, we only have a pointer to the link we want to remove.

If it were singly linked, we would need to somehow find the predecessor (walking the list from the start in O(n)) so that we can change the `next` pointer there.